contrib/hyperscan/matching/input_matchers/source/matcher.cc

```diff
@@ -95,12 +59,88 @@ bool Matcher::match(absl::optional<absl::string_view> input) {
       },
       &matched);
   if (err != HS_SUCCESS && err != HS_SCAN_TERMINATED) {
-    ENVOY_LOG_MISC(error, "unable to scan, error code {}.", err);
+    ENVOY_LOG_MISC(error, "unable to scan, error code {}", err);
   }
 
   return matched;
 }
 
+std::string Matcher::replaceAll(absl::string_view value, absl::string_view substitution) const {
+  // Find matched pair.
+  using Found = std::pair<unsigned long long, unsigned long long>;
+  std::vector<Found> founds;
+  hs_scratch_t* scratch_ = tls_->get()->scratch_;
+  hs_error_t err = hs_scan(
+      som_database_, value.data(), value.size(), 0, scratch_,
+      [](unsigned int, unsigned long long from, unsigned long long to, unsigned int,
+         void* context) -> int {
+        std::vector<Found>* founds = static_cast<std::vector<Found>*>(context);
+        founds->push_back({from, to});
+
+        // Continue searching.
+        return 0;
+      },
+      &founds);
+  if (err != HS_SUCCESS && err != HS_SCAN_TERMINATED) {
+    ENVOY_LOG_MISC(error, "unable to scan, error code {}", err);
+    return std::string(value);
+  }
+
+  // Sort founds. Make sure the longest length found in the front will appear first.
+  std::sort(founds.begin(), founds.end(), [](Found a, Found b) {
+    if (a.first == b.first) {
+      return a.second > b.second;
+    }
+    return a.first < b.first;
+  });
+
+  // Concatenate string and replace matched pair with substitution.
+  std::string result;
+  unsigned long long pos = 0;
+  for (auto& found : founds) {
+    if (found.first < pos) {
+      continue;
+    }
+
+    result += std::string(value.substr(pos, found.first - pos));
```

Comment:
`absl::StrCat(&result, value.substr(pos, found.first - pos), substitution);`

That will save a bunch of extraneous string-copying.

Or, even better, make result be a `vector<string_view>` and at the end write `return absl::StrJoin(result, "");`

Reply:
That's cool! I will fix that.